app/components/ui/BookmarkModal/BookmarkModal.tsx

```diff
@@ -0,0 +1,216 @@
+
+export const BookmarkModal = ({
+  isOpen,
+  setIsOpen,
+  hit,
+}: {
+  isOpen: boolean;
+  setIsOpen: (val: boolean) => void;
+  hit: SearchHit;
+}) => {
+  // TODO update any types as needed when API work is done
+  interface Folder {
+    id?: number;
+    name: string;
+  }
+
+  // TODO: If bookmark already exists, use bookmark name; otherwise,
+  // default to service/resource name. This should be updated with the correct
+  // bookmark name prop when the API is operational
+  const initialName = hit.name;
+  // TODO: Once the service/resource can have an associated bookmark model, we can check
+  // if the bookmark already exists and update various UI on the modal accordingly
+  const newBookmark = false;
+
+  const [folderOptions, setFolderOptions] = useState<Folder[]>([]);
+  const [selectedFolder, setSelectedFolder] = useState<Folder | null>(null);
+  const [newFolder, setNewFolder] = useState("");
+  const [expandFolders, setExpandFolders] = useState(false);
+  const [bookmarkName, setBookmarkName] = useState(initialName);
+  const foldersContainerRef = useRef(null);
+
+  useEffect(() => {
+    // TODO: fetch bookmark folders when the API endpoint is ready. The setTimeout function
+    // is a mock request for the purposes of development.
+    const mockResponse = new Promise((resolve) => {
+      setTimeout(() => {
+        const mockFolders = [
+          { id: 1, name: "Cory's Service List", bookmarks: [] },
+          { id: 2, name: "Fred's Service List", bookmarks: [] },
+          { id: 3, name: "Fredd's Service List", bookmarks: [] },
+          { id: 4, name: "Frewwd's Service List", bookmarks: [] },
+          { id: 5, name: "Frqed's Service List", bookmarks: [] },
+          { id: 8, name: "Frrred's Service List", bookmarks: [] },
+          { id: 9, name: "Fre333d's Service List", bookmarks: [] },
+        ];
+        resolve(mockFolders);
+      }, 500);
+    });
+
+    mockResponse.then((value) => {
+      const response = value as Folder[];
+      setFolderOptions(response);
+    });
+  }, []);
+
+  const handleDropdownBlur = (
+    event: React.FocusEvent<HTMLDivElement, Element>
+  ) => {
+    // Closes the dropdown unless the newly focused element is a child of the dropdown
+    const newFocus = event.relatedTarget;
+    const folderElement = foldersContainerRef.current as unknown as Element;
+    if (newFocus && !folderElement.contains(newFocus)) {
+      setExpandFolders(false);
+    }
+  };
+
+  const createBookmark = () => {
+    // TODO: when bookmarks/folder API is done, fix this up
+    if (!selectedFolder || !bookmarkName) return;
+    const isNewFolder = !selectedFolder.id;
+    if (isNewFolder) {
+      // Create new folder then create new bookmark
+    } else {
+      // Just create new bookmark: pass service/resource ID and folder ID, etc.
+    }
+  };
+
+  const setNewFolderName = (name: string) => {
+    setNewFolder(name);
+    if (name.length) {
+      setSelectedFolder({ id: -1, name });
```

Comment:
Now that you've made the `id` field optional, you should just be able to omit the `id: -1` here.